### DISCUSSION TITLE
I want my fleet to service 10 green stops and 20 red stops. How?

### DISCUSSION THREAD

**Discussion by hklarner** [google/or-tools#3379] [Routing (and legacy CP) questions]
I have a routing problem with more stops than can be serviced in one day. There are two types of stops, red stops and green stops. The vehicles must collaborate to service a fixed number of red stops and a fixed number of green stops in total. Say there are 200 red stops and 100 green stops on the map. I have 5 vehicles and together they should service at least 20 red stops and at least 10 green stops.

How can this be formulated?

Can I add a constraint on the sum of all cumulative variables? If so, I could create a red dimension and a green dimension and add a constraint that says that the sum of cumulative vars of the red dimension at the tour end nodes is greater than 20 (and similar for the green dimension).

Or should I somehow tweak the drop penalties? Currently they are constant across all stops. Every stop has a drop penalty because I have far more stops than vehicles to service them.

Any thoughts?

**Comment by jmarca** [comment]
Try it the way you suggested and see if it works.  Maybe something like:

```python
solver = routing.solver()
reds = []
blues = []
for veh in range(manager.GetNumberOfVehicles()):
    end_index = routing.End(veh)
    reds.append(red_dim.CumulVar(end_index))
    blues.append(blue_dim.CumulVar(end_index))
    
solver.Add(solver.Sum(reds) >= red_goal)
solver.Add(solver.Sum(blues) >= blue_goal)
```

My guess is you will have problems with that if it is difficult for the solver to find a valid initial solution.  The problem is the hard lower bound.  The solver might need to go to a solution that violates that constraint in order to find a better solution. In theory the soft lower bounds can work around this, but in practice they also have problems, I've found.  But in this case, a soft lower bound type constraint won't work, because they only apply to dimensions.  I can't see an easy way to set up a dimension that adds up all the reds, all the blues.

Another way is to create a penalty if the total is not met, using a conditional maybe. Like make a cheap second to last node, and an expensive one for each vehicle, require that one of the two be served, require that the second to last node can be reached by any node, but can only transition to the final depot node, and then use a conditional...if the red or blue sums are not met, then the cheap node is forbidden (remove from Next() set).

But that is complicated.  Try the easy way first.  You have enough stops it might be easy for the solver to always find a valid solution.

(Hmm, sorry, I just noticed I switched green to blue.  I was thinking of the old red bus, blue bus issue with discrete choice modeling.)

**Comment by Mizux** [comment]
Food for though:
Could be "fun" to model it using CP-SAT and kind of https://github.com/google/or-tools/blob/main/examples/python/prize_collecting_vrp_sat.py
* Just replace VISIT_VALUE by RED_VALUE/GREEN_VALUE aka 0/1 arrays
* Add constraint that some of red and green meet the goal kind  of
```python
model.Add(
 sum(obj_vars[i] * red_coeffs[i] for i in range(len(obj_vars))) >= red_goal
model.Add(
 sum(obj_vars[i] * gree_coeffs[i] for i in range(len(obj_vars))) >= green_goal
```
note: obj_vars is a array of "is_visited" booleanVar...
* Change objective to minimize the sum of visited arcs aka still find the shortest multi-path

then compare result against the routing model :)
